lib/srv/usermgmt.go

```diff
@@ -0,0 +1,221 @@
+	// users that failed to delete
+	UserCleanup()
+	// Shutdown cancels the UserCleanup loop
+	Shutdown()
+}
+
+type HostUserManagment struct {
+	backend HostUsersBackend
+	ctx     context.Context
+	cancel  context.CancelFunc
+}
+
+var _ HostUsers = &HostUserManagment{}
+
+// CreateUser creates a temporary Teleport user in the TeleportServiceGroup
+func (u *HostUserManagment) CreateUser(name string, ui *services.HostUsersInfo) (io.Closer, error) {
+	tempUser, err := u.backend.Lookup(name)
+	if err != nil && err != user.UnknownUserError(name) {
+		return nil, trace.Wrap(err)
+	}
+	if tempUser != nil {
+		// try to delete even if the user already exists as only users
+		// in the teleport-system group will be deleted and this way
+		// if a user creates multiple sessions the account will
+		// succeed in deletion
+		return &userCloser{
+			username: name,
+			users:    u,
+			backend:  u.backend,
+		}, trace.AlreadyExists("User already exists")
+	}
+
+	groups := append(ui.Groups, types.TeleportServiceGroup)
+	var errs []error
+	for _, group := range groups {
+		if err := u.createGroupIfNotExist(group); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+	}
+	if err := trace.NewAggregate(errs...); err != nil {
+		return nil, trace.WrapWithMessage(err, "error while creating groups")
+	}
+
+	err = u.backend.CreateUser(name, groups)
+	if err != nil && !trace.IsAlreadyExists(err) {
+		return nil, trace.WrapWithMessage(err, "error while creating user")
+	}
+	return &userCloser{
+		username: name,
+		users:    u,
+		backend:  u.backend,
+	}, nil
+}
+
+func (u *HostUserManagment) createGroupIfNotExist(group string) error {
+	_, err := u.backend.LookupGroup(group)
+	if err != nil && err != user.UnknownGroupError(group) {
+		return trace.Wrap(err)
+	}
+	err = u.backend.CreateGroup(group)
+	if trace.IsAlreadyExists(err) {
+		return nil
+	}
+	return trace.Wrap(err)
+}
+
+// DeleteAllUsers deletes all host users in the teleport service group.
+func (u *HostUserManagment) DeleteAllUsers() error {
+	users, err := u.backend.GetAllUsers()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	teleportGroup, err := u.backend.LookupGroup(types.TeleportServiceGroup)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	var errs []error
+	for _, name := range users {
+		errs = append(errs, u.DeleteUser(name, teleportGroup.Gid))
+	}
+	return trace.NewAggregate(errs...)
+}
+
+// DeleteUser deletes the user only if they are
+// present in the specified group
+func (u *HostUserManagment) DeleteUser(username string, gid string) error {
+	tempUser, err := u.backend.Lookup(username)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	ids, err := u.backend.UserGIDs(tempUser)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for _, id := range ids {
+		if id == gid {
+			err := u.backend.DeleteUser(username)
+			if errors.Is(err, ErrUserLoggedIn) {
+				log.Warnf("Not deleting user %q, user has another session, or running process", username)
```

Comment:
Will it keep printing this warning over and over while the session is active? I would move to debug if so.